ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -0,0 +1,265 @@
+    
+    s_virtualizationRoots[rootIndex].providerUserClient = nullptr;
+    vnode_put(s_virtualizationRoots[rootIndex].rootVNode);
+}
+
+
+- (void)testRegisterProviderForPath_InsertionSucceeded
+{
+    const char* path = "/Users/test/code/Repo";
+    
+    VnodePointer vnode = vnode::Create(self->testMountPoint, path, VDIR);
+    
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, 0);
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(result.root));
+    if (VirtualizationRoot_IsValidRootHandle(result.root))
+    {
+        XCTAssertEqual(s_virtualizationRoots[result.root].providerUserClient, &self->dummyClient);
+
+        XCTAssertTrue(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+        XCTAssertTrue(MockCalls::DidCallFunction(ProviderUserClient_UpdatePathProperty));
+    
+        s_virtualizationRoots[result.root].providerUserClient = nullptr;
+        vnode_put(s_virtualizationRoots[result.root].rootVNode);
+    }
+}
+
+- (void)testRegisterProviderForPath_ArrayFull
+{
+    const char* path = "/Users/test/code/Repo";
+    
+    VnodePointer vnode = vnode::Create(self->testMountPoint, path, VDIR);
+    
+    Memory_FreeArray(s_virtualizationRoots, s_maxVirtualizationRoots);
+    s_maxVirtualizationRoots = INT16_MAX + 1;
+    s_virtualizationRoots = Memory_AllocArray<VirtualizationRoot>(INT16_MAX + 1);
+    memset(s_virtualizationRoots, 0, s_maxVirtualizationRoots * sizeof(s_virtualizationRoots[0]));
+    
+    for (uint32_t i = 0; i < s_maxVirtualizationRoots; ++i)
+    {
+        s_virtualizationRoots[i].inUse = true;
+    }
+    
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&self->dummyClient, self->dummyClientPid, path);
+    XCTAssertEqual(result.error, ENOMEM);
+    XCTAssertFalse(VirtualizationRoot_IsValidRootHandle(result.root));
+
+    XCTAssertFalse(MockCalls::DidCallFunction(vfs_setauthcache_ttl));
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderUserClient_UpdatePathProperty));
+}
+
+
+- (void)testRegisterProviderForPath_ExistingRecycledRoot
+{
+    const char* path = "/Users/test/code/Repo";
+    
+    VnodePointer oldVnode = vnode::Create(self->testMountPoint, path, VDIR);
+    
+    const VirtualizationRootHandle rootIndex = 2;
+    XCTAssertLessThan(rootIndex, s_maxVirtualizationRoots);
+    
+    s_virtualizationRoots[rootIndex].inUse = true;
+    s_virtualizationRoots[rootIndex].rootVNode = oldVnode.get();
+    s_virtualizationRoots[rootIndex].rootVNodeVid = oldVnode->GetVid();
+    s_virtualizationRoots[rootIndex].rootFsid = self->testMountPoint->GetFsid();
+    uint64_t inode = oldVnode->GetInode();
+    s_virtualizationRoots[rootIndex].rootInode = inode;
+    
+    oldVnode->StartRecycling();
+    
+    VnodePointer newVnode = vnode::Create(self->testMountPoint, path, VDIR, oldVnode->GetInode());
```

Comment:
How does the recycling of `oldVnode` have an impact on `newVnode`?

Reply:
It doesn't really, but we definitely don't want to have 2 vnodes for the same path, so the old one must be recycled first.